zebra/dplane_fpm_pb.c

```diff
@@ -0,0 +1,993 @@
+		_Atomic uint32_t obuf_bytes;
+		/* Output buffer peak usage. */
+		_Atomic uint32_t obuf_peak;
+
+		/* Amount of connection closes. */
+		_Atomic uint32_t connection_closes;
+		/* Amount of connection errors. */
+		_Atomic uint32_t connection_errors;
+
+		/* Amount of user configurations: FPE_RECONNECT. */
+		_Atomic uint32_t user_configures;
+		/* Amount of user disable requests: FPE_DISABLE. */
+		_Atomic uint32_t user_disables;
+
+		/* Amount of data plane context processed. */
+		_Atomic uint32_t dplane_contexts;
+		/* Amount of data plane contexts enqueued. */
+		_Atomic uint32_t ctxqueue_len;
+		/* Peak amount of data plane contexts enqueued. */
+		_Atomic uint32_t ctxqueue_len_peak;
+
+		/* Amount of buffer full events. */
+		_Atomic uint32_t buffer_full;
+	} counters;
+
+
+} * gfpc;
+
+enum fpm_pb_events {
+	/* Ask for FPM to reconnect the external server. */
+	FPE_RECONNECT,
+	/* Disable FPM. */
+	FPE_DISABLE,
+	/* Reconnect request by our own code to avoid races. */
+	FPE_INTERNAL_RECONNECT,
+	/* Reset counters. */
+	FPE_RESET_COUNTERS,
+};
+
+#define FPM_RECONNECT(fpc)                                                     \
+	event_add_event((fpc)->fthread->master, fpm_process_event, (fpc),      \
+			FPE_INTERNAL_RECONNECT, &(fpc)->t_event)
+
+#define WALK_FINISH(fpc, ev)                                                   \
+	event_add_event((fpc)->fthread->master, fpm_process_event, (fpc),      \
+			(ev), NULL)
+
+/*
+ * Prototypes.
+ */
+static void fpm_reconnect(struct fpm_pb_ctx *fpc);
+static void fpm_connect(struct event *t);
+static void fpm_process_event(struct event *t);
+static void fpm_process_queue(struct event *t);
+static int fpm_pb_process(struct zebra_dplane_provider *prov);
+static ssize_t protobuf_msg_encode(struct zebra_dplane_ctx *ctx, uint8_t *data,
+				   size_t datalen);
+static int fpm_pb_enqueue(struct fpm_pb_ctx *fpc, struct zebra_dplane_ctx *ctx);
+static Fpm__Message *create_fpm_message(qpb_allocator_t *allocator,
+					struct zebra_dplane_ctx *ctx);
+static Fpm__AddRoute *create_add_route_message(qpb_allocator_t *allocator,
+					       struct zebra_dplane_ctx *ctx);
+static Qpb__L3Prefix *route_prefix_create(qpb_allocator_t *allocator,
+					  struct prefix *p);
+static Fpm__RouteKey *route_key_create(qpb_allocator_t *allocator,
+				       struct prefix *prefix);
+
+
+/*
+ * CLI.
+ */
+#define FPM_STR "Forwarding Plane Manager configuration\n"
+
+DEFUN(fpm_set_address, fpm_set_address_cmd,
+      "fpm address <A.B.C.D|X:X::X:X> [port (1-65535)]",
+      FPM_STR "FPM remote listening server address\n"
+	      "Remote IPv4 FPM server\n"
+	      "Remote IPv6 FPM server\n"
+	      "FPM remote listening server port\n"
+	      "Remote FPM server port\n")
+{
+	struct sockaddr_in *sin;
+	struct sockaddr_in6 *sin6;
+	uint16_t port = 0;
+	uint8_t naddr[INET6_BUFSIZ];
+
+	if (argc == 5)
+		port = strtol(argv[4]->arg, NULL, 10);
+
+	/* Handle IPv4 addresses. */
+	if (inet_pton(AF_INET, argv[2]->arg, naddr) == 1) {
```

Comment:
No need to handle this way. `str2sockunion()` should do this already and you will know `su->sa.sa_family == AF_INET`